Add direct conversion from Gnostic v2 types to spec.Swagger

alexzielenski opened the pull request:
This PR continues conversation from https://github.com/kubernetes/kube-openapi/pull/279 and privately with @liggitt wherein we decided the best path to convert gnostic to kube-openapi was to write a direct conversion.

This PR adds a direct conversion from gnostic openapiv2 types to kube-openapi's spec.Swagger. Below are benchmark results from the same benchmark used in #279:

```
BenchmarkGnosticConversion/json->swagger
BenchmarkGnosticConversion/json->swagger-8                  2     537177100 ns/op    95737132 B/op    1381637 allocs/op
BenchmarkGnosticConversion/swagger->json
BenchmarkGnosticConversion/swagger->json-8                  9     125024798 ns/op    71075921 B/op     274071 allocs/op
BenchmarkGnosticConversion/json->gnostic
BenchmarkGnosticConversion/json->gnostic-8                  5     229392000 ns/op    81446859 B/op    1248395 allocs/op
BenchmarkGnosticConversion/gnostic->pb
BenchmarkGnosticConversion/gnostic->pb-8                  100      11095008 ns/op     2899968 B/op          1 allocs/op
BenchmarkGnosticConversion/gnostic->yaml
BenchmarkGnosticConversion/gnostic->yaml-8                 42      28717820 ns/op    32855169 B/op     264562 allocs/op

BenchmarkGnosticConversion/pb->gnostic
BenchmarkGnosticConversion/pb->gnostic-8                   97      13012220 ns/op     9480701 B/op     123829 allocs/op
BenchmarkGnosticConversion/gnostic->swagger
BenchmarkGnosticConversion/gnostic->swagger-8              50      25910391 ns/op    22287938 B/op     164414 allocs/op
```

pb->gnostic->swagger using this method takes ~39ms

this compares favorably to pb->gnostic->yaml->swagger from the last PR in ~97ms

which compared favorably to pb->gnostic->yaml->json->swagger, today's existing solution, which clocked this benchmark at ~628ms

Caveats:

* Fields like `Maximum`, `Minimum`, `MaxItems`, etc in gnostic are not implemented as pointers. This means the conversion has no way to differentiate between one of these being `0` vs being unused
* Kube-Openapi Swagger type has a number of fields added in OpenAPI v3, despite the fact it represents an OpenAPI v2 object. This fields cannot be populated by the unmarshaler since gnostic errors/ignores unrecognized keys. Thus, if one is roundtripping with gnostic via `kube->json->gnostic->kube`, those fields would be missing, since `json->gnostic` would ignore them.
* spec.SchemaOrArray in gnostic's implementation only scans for one element, so successive schemas are ignored
* A number of our kube-openapi Swagger types do not have `VendorExtensible` embedded even when OpenAPI spec permits "x-" extensions. This means any extensions used in gnostic for a few types are not carried over. This can be corrected in another PR
* Response.Description and Parameter.Description are missing from our type definitions despite being available in openapi v2, so the information is dropped when converted from gnostic. This can be corrected in another PR.

For the use case of kubernetes' swagger.json, most of these caveats do not apply:
* The Maximum/Minimum/etc. fields. May be an issue. I have not looked into whether in the OpenAPI spec these values being 0 means "unset" like gnostic treats them.
* The openapi v3 validation fields like anyOf, not, etc. are stripped from the kubernetes openapi v2 published description. (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/skeleton.go)
* Practical (and kubernetes') usages of Schema.Items (only usage of spec.SchemaOrArray) always have one element. This is because JSON schema spec does not specify it as an array in more recent versions. (`prefixItems` is used as the array version of old `Items`): http://json-schema.org/draft/2020-12/json-schema-core.html#items
* This is only a problem if the information existed in the first place. Since kubernetes use case is kube->gnostic->kube, there is no risk of losing fields that are available in gnostic but not in kube-openapi.
* MIssing fields can be added. But also see above point.

@natasha41575
/cc @apelisse

liggitt commented:
numbers look really promising, thanks

is this ready for review? I see a lot of question marks and commented out code

alexzielenski commented:
@liggitt It is ready for review as of my latest force push. You looked too quickly!

apelisse reviewed (commented):
Minor nits, the conversion code seems good to me, especially if it round-trips properly.

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+
+	max := g.GetMaximum()
+	if max != 0 {
+		k.Maximum = &max
```
does this mean `maximum: 0` does not round-trip?

alexzielenski replied:
That is correct. Maximum: 0 does not round trip. I am not sure how to fix this.

apelisse replied:
Yeah, that's not new for clients that download protobuf though. Also, I was talking to @timburks about that, we can probably fix it in openapi v3.0.

alexzielenski replied:
@apelisse can you elaborate on  “that’s not for new clients”

apelisse replied:
Clients that use the protobuf version (basically kubectl AFAIK), already can't distinguish between 0 and unset. So that's not a new behavior. This is only going to affect clients that start doing the conversion from gnostic to swagger, and there are none today. And it's not something we use today in `kubectl explain` for example (we don't take advantage of this information in there, yet).

liggitt replied:
ok

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+		k.MultipleOf = &multOf
+	}
+
+	if anyEnum, ok := g.(GnosticCommonValidationsAnyEnum); ok {
```
simplify to 
```go
switch g := g.(type) {
case GnosticCommonValidationsAnyEnum:
  enums := g.GetEnum()
  ...
case GnosticCommonValidationsNamedEnum:
  enums := g.GetEnum()
  ...
}
```

alexzielenski replied:
corrected

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+	if anyEnum, ok := g.(GnosticCommonValidationsAnyEnum); ok {
+		enums := anyEnum.GetEnum()
+
+		if len(enums) > 0 {
```
is `enum: []` valid or distinct from `enum: null`?

alexzielenski replied:
I've gone through to correct instance where len was checked in case of null, so that an instance of our object is always created if the gnostic version is also non-nil.

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+		return nil
+	}
+
+	// License does not embed to VendorExtensible!
+	// data loss from g.VendorExtension
```
we know there are field mismatches, and we don't expect this mismatched fields to occur in kube openapi we're parsing, but it seems like we should propagate the fact that there was data loss up so it's visible at the top (we can check in tests, and consumers would at least have visibility to warn or handle as they see fit). That would change return signatures to `(ok bool, err error)`, and we could flip `ok` false for data loss scenarios

alexzielenski replied:
i've changed the signature to now return an `ok` result

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+		return err
+	}
+
+	return k.VendorExtensible.FromGnostic(g.VendorExtension)
```
nit: suggest same pattern of 

```
if err := k.VendorExtensible.FromGnostic(g.VendorExtension); err != nil {
  return err
}

return nil
```

alexzielenski replied:
corrected

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+		k.Properties = make(map[string]Schema)
+		for _, namedSchema := range g.Properties.AdditionalProperties {
+			val := &Schema{}
+			if err := val.FromGnostic(namedSchema.Value); err != nil {
```
namedSchema can be null, right?

alexzielenski replied:
null checks in all ranges over loops have been added/verified

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+		if err := k.FromGnostic(schema); err != nil {
+			return nil
+		}
+	}
```
need a default case that either returns an error or a boolean for data loss

alexzielenski replied:
this was actually a mistake to `return nil`. I meant to `return err`. In any case, the boolean has been added, and the erroneous return corrected.

liggitt commented:
Had a few comments:
* zero-value handling for scalars (might be a limitation of gnostic, which is unfortunate... is that causing problems elsewhere?)
* null vs empty handling of lists/maps
* null pointer checks when ranging over lists of pointers
* tracking/propagating whether unexpected data loss occurred

I didn't review the specific gnostic structs carefully to check all the fields were caught... I'm assuming fuzzing will catch simple field copies.

For the interface types we're doing type switches on, do you have the unit test coverage handy to see how well the tests are exercising all the cases?

alexzielenski commented on pkg/validation/spec/gnostic_test.go:
```diff
@@ -0,0 +1,507 @@
+	)
+}
+
+func TestGnosticConversionSmallDeterministic3(t *testing.T) {
+	// A failed case of TestGnosticConversionSmallRandom
+	// which failed during development/testing loop
+	gnosticCommonTest(
+		t,
+		fuzz.
+			NewWithSeed(1646772024).
+			NilChance(0.8).
+			MaxDepth(10).
+			NumElements(1, 2),
+	)
+}
+
+func TestGnosticConversionSmallDeterministic4(t *testing.T) {
+	// A failed case of TestGnosticConversionSmallRandom
+	// which failed during development/testing loop
+	gnosticCommonTest(
+		t,
+		fuzz.
+			NewWithSeed(1646791953).
+			NilChance(0.8).
+			MaxDepth(10).
+			NumElements(1, 2),
+	)
+}
+
+
+func TestGnosticConversionSmallRandom(t *testing.T) {
+	seed := time.Now().Unix()
+	t.Log("Using seed: ", seed)
+
+	gnosticCommonTest(
+		t,
+		fuzz.
+			NewWithSeed(seed).
+			NilChance(0.8).
+			MaxDepth(10).
```
I disagree that the "deterministic" tests add no value. The `SmallRandom` test case does as you described to fuzz for new failing test cases. Once a failing case is discovered by the random fuzzer it is easy to take the seed used to reproduce the failure, fix it, and add a regression test. Do this enough times and you have a corpus of tricky test cases the code is known to have failed in the past against to guard against regression.

To fuzz in a loop 100 times without printing the random seed used for each case offers no value to a developer other than to say "it's broken", since the exact case can't be reproduced.

alexzielenski replied:
> even changing parameters like this. I don't think the "size" really has an impact for these.

I was forced to change the parameters since the fuzz to generate the object would take too long otherwise. The different "sizes" was useful for me during development to work on smaller cases and then larger and larger. The distinction may not be useful as part of the test sweet but I see no need to remove them.

alexzielenski commented on pkg/validation/spec/gnostic_test.go:
```diff
@@ -0,0 +1,520 @@
+				}
+			}
+		})
+	}
```
done

alexzielenski commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1226 @@
+		}
+	} else if namedEnum, ok := g.(GnosticCommonValidationsNamedEnum); ok {
+		enums := namedEnum.GetEnum()
+		if len(enums) > 0 {
```
corrected. has been changed to a nil check

alexzielenski commented:
@liggitt With my latest updates I think I have addressed most of your comments:

1. Zero-handling: Unfortunately there is no way to differentiate 0 with unspecified here. I also looked into the `protoreflect` api, and [even that has logic](https://github.com/protocolbuffers/protobuf-go/blob/master/internal/impl/message_reflect_field.go#L292) for treating `0` as not being present. A shame. I haven't seen mention of this causing problems for anyone else, but I also haven't looked very hard.
2. I've reviewed the creation of lists/maps to make the handling of empty lists/maps consistent: if gnostic has nil, kube-openapi will also use nil. And vice versa.
3. I've also reviewed pointer usage especially regarding ranges over lists/maps of pointers. Appropriate guards are now in place.
4. A new `ok` result has been added to track whether data loss was detected.

I gathered test coverage info for the gnostic tests. `gnostic.go` is currently 68.8% covered. I'm looking into increasing this.

liggitt commented:
> 1. Zero-handling: Unfortunately there is no way to differentiate 0 with unspecified here. I also looked into the `protoreflect` api, and [even that has logic](https://github.com/protocolbuffers/protobuf-go/blob/master/internal/impl/message_reflect_field.go#L292) for treating `0` as not being present. A shame. I haven't seen mention of this causing problems for anyone else, but I also haven't looked very hard.

huh, that's a pretty sharp edge. I can easily imagine someone using `maximum: 0`. not sure what to do about that other than document it, and make sure authoritative validators don't use this method?

> 2. I've reviewed the creation of lists/maps to make the handling of empty lists/maps consistent: if gnostic has nil, kube-openapi will also use nil. And vice versa.

👍 

> 3. I've also reviewed pointer usage especially regarding ranges over lists/maps of pointers. Appropriate guards are now in place.

👍

> 4. A new `ok` result has been added to track whether data loss was detected.

👍

> I gathered test coverage info for the gnostic tests. `gnostic.go` is currently 68.8% covered. I'm looking into increasing this. Upon a quick glance, at least not all of the type cases of Parameter are being exercised.

👍

alexzielenski commented:
With my latest commit all supported gnostic types and significant LOC are now being exercised within the tests.

alexzielenski commented:
/retest

liggitt commented on go.mod:
```diff
@@ -22,6 +22,7 @@ require (
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.5.1
 	golang.org/x/tools v0.1.5 // indirect
+	google.golang.org/protobuf v1.27.1
```
nit, looks like `go mod tidy` is needed (`go test` complained when I ran it locally)

alexzielenski replied:
done

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+
+	if enums != nil {
+		k.Enum = make([]interface{}, len(enums))
+		for i, v := range enums {
```
still a fair number of places we're ranging over pointers that could be nil... I guess we should always check nil and `continue` if nil, rather than accessing fields or calling methods on those things.

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+			} else if !nok {
+				ok = false
+			}
+		} else if nk, err := strconv.Atoi(v.Name); err == nil {
```
if err is non-nil, this should either return an error or at least set `ok` to false (and add a testcase to cover this)

alexzielenski replied:
done

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+				ok = false
+			}
+		} else {
+			for _, v := range g.Items.GetSchema() {
+				schema := Schema{}
+				if nok, err := schema.FromGnostic(v); err != nil {
+					return false, err
+				} else if !nok {
+					ok = false
+				}
+				k.Items.Schemas = append(k.Items.Schemas, schema)
```
this block doesn't have test coverage

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+			ok = false
+		}
+		return ok, nil
+	case *openapi_v2.SecurityDefinitionsItem_Oauth2AccessCodeSecurity:
+		if nok, err := k.SecuritySchemeProps.FromGnostic(s.Oauth2AccessCodeSecurity); err != nil {
+			return false, err
+		} else if !nok {
+			ok = false
+		}
+		if nok, err := k.VendorExtensible.FromGnostic(s.Oauth2AccessCodeSecurity.VendorExtension); err != nil {
+			return false, err
+		} else if !nok {
+			ok = false
+		}
+		return ok, nil
```
this case is missing test coverage

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+	// License does not embed to VendorExtensible!
+	// data loss from g.VendorExtension
+	if len(g.VendorExtension) != 0 {
+		ok = false
```
This data loss code path is missing test coverage. Will probably need a bespoke (non-fuzzed) test constructing a gnostic object we know has data we can't convert, and make sure `!ok` gets plumbed all the way back through to Swagger#FromGnostic correctly.

alexzielenski replied:
added a specific test targeting data loss cases

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+	// ContactInfo does not embed to VendorExtensible!
+	// data loss from g.VendorExtension
+	if len(g.VendorExtension) != 0 {
+		ok = false
```
this data loss code path is missing test coverage

alexzielenski replied:
added a specific test targeting data loss cases

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+
+	// data loss! g.VendorExtension
+	if len(g.VendorExtension) != 0 {
+		ok = false
```
this data loss code path is missing test coverage

alexzielenski replied:
added a specific test targeting data loss cases

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+
+			return ok, nil
+		default:
+			return false, errors.New("unrecognized nobody type for Parameter")
```
```suggestion
			return false, errors.New("unrecognized nonbody type for Parameter")
```

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+			return false, errors.New("unrecognized nobody type for Parameter")
+		}
+	default:
+		return false, errors.New("unrecognized type for Paramter")
```
```suggestion
		return false, errors.New("unrecognized type for Parameter")
```

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+		} else if !nok {
+			ok = false
+		}
+	}
```
still needs a default case that returns an error, right?

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+			ok = false
+		}
+		return ok, nil
+	}
```
needs a default case that returns an error, right?

alexzielenski replied:
yup

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+
+// Swagger Document
+
+func (k *Swagger) FromGnostic(g *openapi_v2.Document) (ok bool, err error) {
```
document the meaning of the ok and err returns (`!ok` means a value we know we can't round-trip was present in the gnostic document, err means an unexpected error occurred converting)

liggitt replied:
also document that zero-value validation constraints like maxItems/maxLength/maxProperties are lost via gnostic

alexzielenski replied:
added

liggitt commented:
looking really solid, thanks. noted a few places we want to make sure we exercise plumbing with tests

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+	return ok, nil
+}
+
+func (k *HeaderProps) FromGnostic(g *openapi_v2.Header) (ok bool, err error) {
```
drop the `ok` from this method, since it's unused

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+	return ok, nil
+}
+
+func (k *SecurityScheme) FromGnostic(g *openapi_v2.SecurityDefinitionsItem) (ok bool, err error) {
```
I think `ok` can be dropped from here once SecuritySchemeProps/VendorExtensible stop returning `ok`

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+
+// SecurityDefinitions
+
+func (k SecurityDefinitions) FromGnostic(g *openapi_v2.SecurityDefinitions) (ok bool, err error) {
```
I think `ok` can be dropped once SecuritySchemeProps stops returning `ok`

liggitt reviewed (commented):
one way to improve coverage of tested code paths is to remove unused ones... I noted several places where we have no use of the `ok` data loss return indicator... trimming that to only places that are propagating errors from License/ContactInfo/ExternalDocumentation will help clarify where we could actually exercise those branches in tests

alexzielenski commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1554 @@
+	GetEnum() []*openapi_v2.Any
+}
+
+func (k *CommonValidations) FromGnostic(g GnosticCommonValidations) (ok bool, err error) {
```
there is one path which sets it to `false`. Below this point by the handling for `Enum` (unsure how to get permalink to line of code).

alexzielenski commented:
/lgtm
/approve

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1496 @@
+
+			k.Security[i] = make(map[string][]string, len(v.AdditionalProperties))
+			converted := k.Security[i]
+			for _, p := range v.AdditionalProperties {
```
check for nil p

alexzielenski replied:
done

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1496 @@
+package spec
```
add to new files:

```
/*
Copyright 2022 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

alexzielenski replied:
done

liggitt reviewed (commented):
a couple last nits, then lgtm

liggitt commented on pkg/validation/spec/gnostic.go:
```diff
@@ -0,0 +1,1514 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package spec
```
sorry, leave a blank line between the comment and the package decl so this doesn't show up in godoc as package docs (same in the other file)

alexzielenski replied:
fixed and noted for future

liggitt commented:
/lgtm
/approve

apelisse commented:
/approve

Thanks, that's awesome! Can't wait to see the opposite direction 😂
